src/OmniSharp.Host/Program.cs

```diff
@@ -18,168 +21,144 @@ namespace OmniSharp
+                    {
+                        System.Threading.ThreadPool.SetMinThreads(MIN_WORKER_THREADS, currentCompletionPortThreads);
+                    }
                 }
-            }
 #endif
 
-            Environment = new OmniSharpEnvironment(applicationRoot, serverPort, hostPID, logLevel, transportType, otherArgs.ToArray());
-
-            var config = new ConfigurationBuilder()
-                .AddCommandLine(new[] { "--server.urls", $"http://{serverInterface}:{serverPort}" });
-
-            // If the --encoding switch was specified, we need to set the InputEncoding and OutputEncoding before
-            // constructing the SharedConsoleWriter. Otherwise, it might be created with the wrong encoding since
-            // it wraps around Console.Out, which gets recreated when OutputEncoding is set.
-            if (transportType == TransportType.Stdio && encoding != null)
-            {
-                Console.InputEncoding = encoding;
-                Console.OutputEncoding = encoding;
+                return Run(args);
             }
-
-            var writer = new SharedConsoleWriter();
-
-            var builder = new WebHostBuilder()
-                .UseConfiguration(config.Build())
-                .UseEnvironment("OmniSharp")
-                .UseStartup(typeof(Startup))
-                .ConfigureServices(serviceCollection =>
-                {
-                    serviceCollection.AddSingleton<IOmniSharpEnvironment>(Environment);
-                    serviceCollection.AddSingleton<ISharedTextWriter>(writer);
-                    serviceCollection.AddSingleton<PluginAssemblies>(new PluginAssemblies(plugins));
-                    serviceCollection.AddSingleton<IAssemblyLoader, AssemblyLoader>();
-                });
-
-            if (transportType == TransportType.Stdio)
-            {
-                builder.UseServer(new StdioServer(Console.In, writer));
-            }
-            else
+            catch (Exception e)
             {
-                builder.UseKestrel();
+                Console.Error.WriteLine(e.GetBaseException().Message);
+                return 0xbad;
             }
+        }
+
+        public static int Run(string[] args)
+        {
+            Console.WriteLine($"OmniSharp: {string.Join(" ", args)}");
 
-            using (var app = builder.Build())
+            // omnisharp.json arguments should not be parsed by the CLI args parser
+            // they will contain "=" so we should filter them out
+            var omnisharpJsonArgs = args.Where(x => x.Contains("="));
+
+            var omnisharpApp = new CommandLineApplication(throwOnUnexpectedArg: false);
+            omnisharpApp.HelpOption("-? | -h | --help");
+
+            var applicationRootOption = omnisharpApp.Option("-s | --solution", "Solution, project file or directory for OmniSharp to point at (defaults to current directory).", CommandOptionType.SingleValue);
```

Comment:
Since we have an opportunity to name the long name, how about something more general. E.g. `--source`